lib/drivers/docker/container.js

```diff
@@ -0,0 +1,360 @@
+  var self = this;
+  this.docker.getContainer(this.name).inspect(function(err, details) {
+    if (err) {
+      self._create();
+    } else {
+      self.c = self.docker.getContainer(details.Id);
+      self.cid = details.Id.slice(0, 12);
+      self.emit('created');
+    }
+  });
+};
+
+Container.prototype._create = function() {
+  var self = this;
+  this.docker.createContainer(this.createOpts, function(err, c) {
+    self.debug('creatd container:', c);
+    if (err) {
+      self.emit('error', err);
+    } else {
+      self.c = c;
+      // similar to git, don't need the whole id
+      self.cid = c.id.slice(0, 12);
+      self.emit('created');
+    }
+  });
+};
+
+Container.prototype.destroy = function(cb) {
+  var self = this;
+  this.debug('Destroying docker/container instance', this.cid);
+  // ensure the master process is properly reported as stopped
+  this.emit('exit', 'container destroyed');
+  this.c.remove({v: true, force: true}, function(err) {
+    if (err) {
+      console.error('pm:docker:container: error removing container:', err);
+    } else {
+      console.log('pm:docker: removed container for %j: %j',
+                  self.instanceId, self.cid);
+    }
+    // lazily cleanup event emitters so that any last words can be heard
+    setTimeout(function() {
+      self.removeAllListeners();
+    }, 5000).unref();
+    if (cb) {
+      setImmediate(cb);
+    }
+  });
+  this.driver =
+    this.docker =
+    this.logger =
+    this.c = null;
+};
+
+Container.prototype.start = function() {
+  this.debug('Docker::Container.start()');
+  var self = this;
+  var opts = {
+    // Paper over a bug in Docker <1.6:
+    // according to the Docker API docs, this is a create option, and start
+    // doesn't even take options, but this seems to be the only way it works.
+    PublishAllPorts: true,
+  };
+  this.c.start(opts, function(err) {
+    if (err && err.reason !== 'container already started') {
+      self.emit('error', err);
+    } else {
+      self.emit('starting');
+    }
+  });
+};
+
+Container.prototype.startSize = function() {
+  if (/cpu/i.test(String(this.startOpts.size))) {
+    return this.driver.CPUS;
+  } else {
+    return this.startOps.size;
+  }
+};
+
+Container.prototype.connectControlChannel = function() {
+  this.debug('Docker::Container.connectControlChannel()');
+  var self = this;
+  this.ctl = new sCtl.Client(this.ports.ctl, onMessage, onMessage, onError);
+  this.ctl._socket.once('connect', function() {
+    self.connected = true;
+    self.emit('connected');
+  });
+  this.ctl._socket.on('error', function(err) {
+    self.debug('socket error on control channel:', err);
+    self.connected = false;
+  });
+
+  function onMessage(req) {
+    setImmediate(self.emit.bind(self, 'request', req));
+  }
+
+  function onError(err) {
+    self.debug('onError: ', err, err.stacktrace);
+    if (err.code === 'ECONNREFUSED') {
+      self.connected = false;
+      self.debug('Retrying connection');
```

Comment:
debugging once every 200ms is painful, could you debug on first error, and debug on connect, so in-between it doesn't log so heavily?